staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -562,6 +563,119 @@ func (s *store) List(ctx context.Context, key string, opts storage.ListOptions,
 	return s.list(ctx, key, opts, listObj, true)
 }
 
+type clientPagingConfig struct {
+	// hasMore guards if we need to set up the remaining configuration
+	hasMore         bool
+	startKey        string
+	resourceVersion int64
+	remainingCnt    int64
+}
+
+func (s *store) paginatedList(ctx context.Context, key, end string, rev int64, fromRev *uint64,
+	v reflect.Value, typeName string, pred storage.SelectionPredicate, newItemFunc func() runtime.Object) (pcfg *clientPagingConfig, err error) {
+	pcfg = &clientPagingConfig{}
+	maximumLimit := int64(500)
+	if !s.pagingEnabled || pred.Limit <= 0 {
+		pred.Limit = math.MaxInt64
+	}
+	remainingLimit := pred.Limit
+
+	// loop until we have filled the requested limit from etcd or there are no more results
+	var lastKey []byte
+	var hasMore bool
+	var getResp *clientv3.GetResponse
+	var numFetched int
+	var numEvald int
+	// Because these metrics are for understanding the costs of handling LIST requests,
+	// get them recorded even in error cases.
+	defer func() {
+		numReturn := v.Len()
+		metrics.RecordStorageListMetrics(s.groupResourceString, numFetched, numEvald, numReturn)
+	}()
+	for {
+		limit := maximumLimit
+		if remainingLimit < limit {
+			limit = remainingLimit
+		}
```

Comment:
I think `appendListItem` only _conditionally_ appends, based on `pred` - and the overall `pred.Limit` should be a count of the successfully matched _results_, not the etcd keyspace searched.  (Did I read the previous code correctly?)

Concretely, I think we should scrap `remainingLimit` and instead just compare v.Len() vs pred.Limit as used elsewhere in this function:

```suggestion
		limit := pred.Limit - int64(v.Len())
		if maximumLimit < limit {
			limit = maximumLimit
		}
```

Reply:
I agree `pred.Limit` should be "a count of the successfully matched results, not the etcd keyspace searched". 

However, considering a query like this `GET /api/v1/events?limit=500&fieldSelector=field=AAA` when kubelet reconnects to APIServer, APIServer literally searches the etcd keyspace in a batch of 500 in current k8s mainline code until match 500 results even if in the middle of the pagination loop, there are a few matched result. 

I think your proposed `pred.Limit` to `limit` translation makes more sense. I will change it.